[FLINK-15259][hive] HiveInspector.toInspectors() should convert Flink…

lirui-apache opened the pull request:
… constant to Hive constant

<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](https://flink.apache.org/contributing/contribute-code.html#open-a-pull-request).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Convert constant Flink object to Hive object before creating Hive's constant object inspectors.


## Brief change log

  - Convert the constant value to Hive object in `HiveInspectors::toInspectors`
  - Implement our own java constant objector inspectors for char, varchar and decimal
  - Add test case


## Verifying this change

New test case

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? NA

lirui-apache commented:
cc @bowenli86 @zjuwangg @JingsongLi @xuefuz

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/conversion/FlinkJavaConstantHiveVarcharObjectInspector.java:
```diff
@@ -0,0 +1,56 @@
+
+	public FlinkJavaConstantHiveVarcharObjectInspector(VarcharTypeInfo varcharTypeInfo, HiveVarchar value) {
+		super(varcharTypeInfo);
+		this.value = value == null || BaseCharUtils.doesPrimitiveMatchTypeParams(value, varcharTypeInfo) ?
```
BaseCharUtils.doesPrimitiveMatchTypeParams() seems check if the value length is equal to the precision, which is valid for CHAR type, but not for VARCHAR.

Also, I saw inconsistency between this one and doesWritableMatchTypeParams().

Can we doubt check on this?

lirui-apache replied:
You're right. We only need to enforce the length if character length exceeds the specified max length.

lirui-apache commented:
Hi @xuefuz , I just updated to address your comment.

JingsongLi commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/client/HiveShimV120.java:
```diff
@@ -100,20 +109,17 @@ public ObjectInspector getObjectInspectorForConstant(PrimitiveTypeInfo primitive
 			case VARCHAR:
-				className = "org.apache.hadoop.hive.serde2.objectinspector.primitive.JavaConstantHiveVarcharObjectInspector";
-				return HiveReflectionUtils.createConstantObjectInspector(className, value);
+				return new FlinkJavaConstantHiveVarcharObjectInspector((VarcharTypeInfo) primitiveTypeInfo, (HiveVarchar) value);
```
Can you explain more why you do this?
What the difference between `JavaConstantHiveVarcharObjectInspector` and `FlinkJavaConstantHiveVarcharObjectInspector`?

lirui-apache commented:
@JingsongLi Let me explain why we need our own implementations for the object inspectors. Hive's Java constant object inspector constructors only take the constant object and not the type info. For types with precision/scale parameters, this causes problem because those parameters need to be stored in the type info. And the missing type info will eventually leads to NPE.
Our implementations also override the `getPrimitiveJavaObject` method because 1) we already know the constant value and should simply return it; 2) Hive's `getPrimitiveJavaObject` doesn't handle writable values and in some cases cause class cast exceptions.

JingsongLi commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/conversion/HiveInspectors.java:
```diff
@@ -118,6 +118,8 @@
 						HiveTypeUtil.toHiveTypeInfo(argTypes[i], false));
 			} else {
 				PrimitiveTypeInfo primitiveTypeInfo = (PrimitiveTypeInfo) HiveTypeUtil.toHiveTypeInfo(argTypes[i], false);
+				constant = getConversion(getObjectInspector(primitiveTypeInfo), argTypes[i].getLogicalType(), hiveShim)
```
This is really a blocker...
Without this line, will throw exception...

lirui-apache commented:
> Consider using `getPrimitiveWritableConstantObjectInspector`?
> I tried, and it worked.

@bowenli86 IIRC, you chose to use java constant object inspectors for constant arguments. Is there a reason that we should prefer java to writable object inspectors?

bowenli86 commented:
> > Consider using `getPrimitiveWritableConstantObjectInspector`?
> > I tried, and it worked.
> 
> @bowenli86 IIRC, you chose to use java constant object inspectors for constant arguments. Is there a reason that we should prefer java to writable object inspectors?

I recall that I might choose java constant oi over writable constant oi because WritableConstantObjectInspector may auto box the java constant which seems not necessary at that time. If auto boxing turns out to be beneficial, we should use it.

BTW, I researched Hive 3.x code base, seems java constant oi is not used anymore, and writable constant oi is the standard now. We should be good to switch.

lirui-apache commented:
@JingsongLi @bowenli86 I have updated to use writable OIs for constant arguments, so that we don't need to our own implementations of OIs. Please note:
1. Non-constant arguments still use java OIs, but I can also update that if you think we should be consistent.
2. Also added test for constant DATE and TIMESTAMP values and realized we need a shim for the `java -> writable` conversion.
3. I haven't removed the shim for getting constant OIs. Hopefully it makes review a little bit easier. Can do that if you think we're moving in the right direction.

JingsongLi commented on flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/module/hive/HiveModuleTest.java:
```diff
@@ -88,4 +94,24 @@ public void testHiveBuiltInFunction() {
+		List<Row> results = TableUtils.collectToList(tEnv.sqlQuery("select concat('an', 'bn')"));
+		assertEquals("[anbn]", results.toString());
+
+		results = TableUtils.collectToList(tEnv.sqlQuery("select concat('ab', cast('cdefghi' as varchar(5)))"));
```
Can you add a test for `null` literal?

lirui-apache replied:
Unfortunately we cannot test `null` literals, because Hive functions consider nulls as non-constant arguments: https://github.com/apache/flink/blob/master/flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/functions/hive/HiveFunction.java#L36

Maybe we should consider all the passed arguments as constant, and don't pass non-constant arguments in the first place. But that'll need to change the interface.

JingsongLi replied:
OK.
Is there any cases to cover null? If not, we can add some whatever whether should be a literal.

JingsongLi replied:
And you can add a TODO to the tests.
About null literal support, we should wait for UDF type reworking.

JingsongLi commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/client/HiveShimV310.java:
```diff
@@ -233,4 +240,23 @@ public LocalDate toFlinkDate(Object hiveDate) {
+	public Writable hivePrimitiveToWritable(Object value) throws CatalogException {
+		try {
+			return super.hivePrimitiveToWritable(value);
+		} catch (CatalogException e) {
```
Any other good way? exception is a little hacky and low performance.

lirui-apache replied:
Maybe I can use optional. Let me have a try

lirui-apache commented:
@JingsongLi PR updated to address your comments.

xuefuz commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/table/catalog/hive/client/HiveShim.java:
```diff
@@ -232,4 +233,10 @@ SimpleGenericUDAFParameterInfo createUDAFParameterInfo(ObjectInspector[] params,
+	 * Converts a Hive primitive java object to corresponding Writable object. Throws CatalogException if the conversion
+	 * is not supported.
+	 */
+	Writable hivePrimitiveToWritable(Object value) throws CatalogException;
```
Throwing a catalog ex doesn't seem very intuitive, even though CatalogEx is also a runtime exception. Maybe just remove CatalogEx from the signature and throw a runtime exception when problem occurs.

lirui-apache replied:
I'll change to throw `FlinkHiveException`. It's also a runtime exception but more specific to Hive connector.

xuefuz commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/GenerateUtils.scala:
```diff
@@ -375,7 +375,7 @@ object GenerateUtils {
            """.stripMargin
         ctx.addReusableMember(fieldTimestamp)
-        generateNonNullLiteral(literalType, fieldTerm, literalType)
+        generateNonNullLiteral(literalType, fieldTerm, ts)
```
What's this change for?

JingsongLi commented on flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/GenerateUtils.scala:
```diff
@@ -375,7 +375,7 @@ object GenerateUtils {
            """.stripMargin
         ctx.addReusableMember(fieldTimestamp)
-        generateNonNullLiteral(literalType, fieldTerm, literalType)
+        generateNonNullLiteral(literalType, fieldTerm, ts)
```
Don't mix the planner's single line modification with hive's modification, and there is no test.
You should add a single commit to explain why we need it and what is this bug.

lirui-apache replied:
@xuefuz @JingsongLi This is a bug in planner when generating timestamp literals: we pass the type instead of the value as the literal. The added test exposes the issue. Do you think this change needs a dedicated PR?

JingsongLi replied:
I think we don't need to dedicated PR for little changes, but we should a dedicated commit at least.

JingsongLi replied:
And I remember this literal value is internal object, it should be `SqlTimestamp`.

lirui-apache replied:
Right, `ts` is a `SqlTimestamp `. I'll squash and make a separate commit for this.

lirui-apache replied:
Now this change is moved to b5f0206

JingsongLi commented:
@lirui-apache Can you organize the format of commits:
`[FLINK-15259][hive] HiveInspector.toInspectors() should convert Flink constant to Hive constant`
`[hotfix][table-planner-blink] Fix generating timestamp literal`

JingsongLi commented:
travis passed in: https://travis-ci.org/lirui-apache/flink/builds/631257572

lirui-apache commented:
> @lirui-apache Can you organize the format of commits:
> `[FLINK-15259][hive] HiveInspector.toInspectors() should convert Flink constant to Hive constant`
> `[hotfix][table-planner-blink] Fix generating timestamp literal`

@JingsongLi Done.
